src/Adapters/EventSource.Netstandard13/EventSourceTelemetryModule.cs

```diff
@@ -218,5 +229,65 @@ private void EnableAsNecessary(EventSource eventSource)
                 }
             }
         }
+
+        /// <summary>
+        /// Returns true when eventSource satisfied the rule; false otherwise. Returns false when either is null.
+        /// </summary>
+        /// <param name="eventSource">The target event source.</param>
+        /// <param name="rule">The naming rule to be used for matching.</param>
+        private bool IsEventSourceNameMatch(EventSource eventSource, EventSourceNamingMatchRuleBase rule)
+        {
+            if (rule == null || eventSource == null)
+            {
+                return false;
+            }
+
+            if (!rule.PrefixMatch)
+            {
+                return string.Equals(eventSource.Name, rule.Name, StringComparison.Ordinal);
+            }
+            else
+            {
+                return IsEventSourceNameMatch(eventSource?.Name, rule?.Name);
+            }
+        }
+
+        /// <summary>
+        /// Returns true when eventSourceName satisfied the rule; false otherwise. Returns false when either is null.
+        /// </summary>
+        /// <param name="eventSourceName"></param>
+        /// <param name="rule"></param>
+        /// <returns></returns>
+        internal bool IsEventSourceNameMatch(string eventSourceName, string rule)
+        {
+            if (string.IsNullOrEmpty(eventSourceName) || string.IsNullOrEmpty(rule))
+            {
+                return false;
+            }
+
+            return eventSourceName.StartsWith(rule);
+        }
+
+        /// <summary>
+        /// When the event comes from a EventSource that matches the rule of DisabledSource, it should be dropped, unless the event source name matches one of the 
+        /// name exactly in the enabling rule.
+        /// </summary>
+        /// <param name="eventData">The event data to test.</param>
+        private bool IsDroppingEvent(EventWrittenEventArgs eventData)
+        {
+            string eventSourceName = eventData?.EventSource?.Name;
+            if (string.IsNullOrEmpty(eventSourceName))
+            {
+                throw new ArgumentNullException(nameof(eventSourceName));
+            }
+
+            // Do NOT drop the event when it matches one of the request by name exactly.
+            if (Sources.Any(request => string.Equals(request.Name, eventSourceName, StringComparison.Ordinal)))
```

Comment:
so you check for equality first and then iterate over the list again? I think you can either split into two lists or simply call `IsEventSourceNameMatch` in one go.

Especially keep in mind that the default config only has one namespace which is never be match exactly

Reply:
Hey, Sorry, I am not sure I got what you mean here. But I think the concern here's performance.

Firstly, in the caller, this logic will be bypassed if there's no item in <DisabledSources>.
Secondly, in the default configuration, since there are no <DisabledSources>, we won't run into this code path.

Based on those conditions, here it is checking against two different lists in terms of <Sources> or <DisabledSources>. Here's the step:
Whenever an event comes in:
1. Check if the event's source belongs to the <Sources> list. If there is a **exact** match, do NOT drop the event - as the exact match in <Sources> always take priority.

2. Otherwise, see if the event's source matches the rule in <DisabledSources>, **exact or prefix match** by the rule. If there is a match, drop the event, else, keep it.

Does this make sense? Or do I missed something? Thanks!

Reply:
My understanding was the you'll include `Microsoft.ApplicaitonInsights` in the list of disabled sources by default as it causes the problem. If so - the list will never be empty and first loop will never result in a match. So I was suggesting to not run every matching rule twice - one for exact match and one for prefix match and do prefix match in the main loop.

Reply:
@SergeyKanzhelev,
I don't think that will be the case if I understand the problem correctly.
My understanding is in most cases, there's no need to set DisabledSource at all unless there's event source starts to cause dead-loop.
@karolz-ms, could you please confirm?

Reply:
@xiaomi7732 correct, by default I would expect the list of disabled sources to be empty

Reply:
Wait. Isn't it the case that `Microsoft.ApplicationInsights` listener will cause problem all the time? And we do enable `Microsoft.ApplicationInsights` listener by default in Application Insights SDK

Reply:
No. In the report that originated this PR it was the Microsoft-ApplicationInsights-Data EventSource (a.k.a. RichPayloadEventSource) that triggered this, but typically either this event source is not used, or the buggy codepath in .NET EventSource/EventListener is not hit, and everything works.

So we are trying to be proactive here and give people a workaround *just in case they need it*.

That plus enabling or disabling sources based on prefix matching can cut the amount of configuration goo quite significantly, so it is a useful feature regardless of the .NET FX bug.

Reply:
ok. If config size a main concern - we can have a default list in code. If we know for sure it may happen and we know that this eventsource will never be listened from this telemetry module - it worth adding the default config.

This is how to initialize default list in code so it can be overridden from config or code later on: https://github.com/Microsoft/ApplicationInsights-dotnet-server/blob/ecd7723b6de2a75530dc18741766825d6f5e42f5/Src/PerformanceCollector/Perf.Shared.NetFull/PerformanceCollectorModule.cs#L104-L111

Reply:
@SergeyKanzhelev, thanks for the reference.

Here's my opinion:
1. I don't think the config size is the main concern.
2. We don't know which event source will be the one to cause the trouble until runtime - calling write event on dynamic EventSource basically will broadcast the event data to all the event source listeners enabled by the system (It doesn't matter if we registered to listen to that event source or not in this case) - the .net fx bug. Among those event sources, if there is one that will loop back to the code to trigger the same dynamic event, directly, or indirectly, a dead loop will be formed. Having the <DisabledSources /> will allow the user to break the loop in that scenario.
3. That said, I don't think we are failing by default. In most cases, the user shall not know the existence of <DisabledSources /> at all.
4. It is not a good idea to put anything into <DisabledSources /> by default because once there's anything in the collection, there will be performance degrade.

Reply:
My problem with this that if `Microsoft-ApplicationInsights-Data` enabled it guaranteed to be called as a result of ANY telemetry item reported by this telemetry module. Thus it is almost inevitably fall into this problem. Or I'm missing something?

 One potential solution will be to have some optimized check specifically for this event source.

Reply:
@SergeyKanzhelev the problem only manifested itself when a combination of EventFlow + ApplicationInsights + NLog was used, please refer to original issue https://github.com/Azure/diagnostics-eventflow/issues/140 . It does not happen when only EventSourceTelemetryModule with the rest of AI SDK is being used, otherwise EventSourceTelemetryModule would never work for anybody and the tests verifying its functionality would be failing all the time.

BUT

The problem was traced down to a bug in EventListener infrastructure https://github.com/dotnet/coreclr/issues/14434 I haven't debugged deeply enough into EventFlow + AI SDK + NLog combination to understand why this combination fails vs. EventFlow + AI SDK or EventSourceTelemetryModule + AI SDK work. But because I was able to reproduce the EventListener problem with a synthetic repro that has no EventFlow, AI SDK, or NLog,  we want to be proactive about giving EventSourceTelemetryModule users a workaround in case they hit it.

Since typical library combinations work, I think it is not necessary to disable the Microsoft-ApplicationInsights-Data EventSource by default.

Does this make sense now?

Reply:
Ok, I trust your judgment here.